File: src/main/java/com/nido/api/shared/model/MeasurementUnit.java
package com.nido.api.shared.model;

import java.math.BigDecimal;

public enum MeasurementUnit {
    GRAM(UnitFamily.MASS, BigDecimal.ONE),
    KILOGRAM(UnitFamily.MASS, BigDecimal.valueOf(1000)),
    MILLILITER(UnitFamily.VOLUME, BigDecimal.ONE),
    CENTILITER(UnitFamily.VOLUME, BigDecimal.TEN),
    LITER(UnitFamily.VOLUME, BigDecimal.valueOf(1000)),
    PIECE(UnitFamily.COUNT, BigDecimal.ONE),
    SLICE(UnitFamily.COUNT, BigDecimal.ONE),
    TABLESPOON(UnitFamily.COUNT, BigDecimal.ONE),
    TEASPOON(UnitFamily.COUNT, BigDecimal.ONE),
    PINCH(UnitFamily.COUNT, BigDecimal.ONE),
    SACHET(UnitFamily.COUNT, BigDecimal.ONE);

    private final UnitFamily family;
    private final BigDecimal factorToBaseUnit;

    MeasurementUnit(UnitFamily family, BigDecimal factorToBaseUnit) {
        this.family = family;
        this.factorToBaseUnit = factorToBaseUnit;
    }

    public UnitFamily family() {
        return family;
    }

    public BigDecimal factorToBaseUnit() {
        return factorToBaseUnit;
    }

    /**
     * Converts a quantity in this unit to the family's base unit (gram for MASS,
     * milliliter for VOLUME). Meaningless across families — callers only ever use
     * this to compare/sum quantities already known to share a family.
     */
    public BigDecimal toBaseUnits(BigDecimal quantity) {
        return quantity.multiply(factorToBaseUnit);
    }
}

File: src/main/frontend/src/shared/lib/measurementUnit.ts
export type MeasurementUnit =
  | 'GRAM' | 'KILOGRAM' | 'MILLILITER' | 'CENTILITER' | 'LITER'
  | 'PIECE' | 'SLICE' | 'TABLESPOON' | 'TEASPOON' | 'PINCH' | 'SACHET'

export const MEASUREMENT_UNITS: MeasurementUnit[] = [
  'GRAM', 'KILOGRAM', 'MILLILITER', 'CENTILITER', 'LITER',
  'PIECE', 'SLICE', 'TABLESPOON', 'TEASPOON', 'PINCH', 'SACHET',
]

/** Translation keys within the `common` i18n namespace — used by any consumer that can't reach into a page-private locale file (e.g. the shopping-list entity/feature). */
export const MEASUREMENT_UNIT_LABEL_KEY: Record<MeasurementUnit, string> = {
  GRAM: 'unit.GRAM',
  KILOGRAM: 'unit.KILOGRAM',
  MILLILITER: 'unit.MILLILITER',
  CENTILITER: 'unit.CENTILITER',
  LITER: 'unit.LITER',
  PIECE: 'unit.PIECE',
  SLICE: 'unit.SLICE',
  TABLESPOON: 'unit.TABLESPOON',
  TEASPOON: 'unit.TEASPOON',
  PINCH: 'unit.PINCH',
  SACHET: 'unit.SACHET',
}
